pkg/apis/batch/validation/validation_test.go

```diff
@@ -1283,6 +1283,114 @@ func TestValidateJobUpdate(t *testing.T) {
 				AllowMutableSchedulingDirectives: true,
 			},
 		},
+		"update completions and parallelism to same value is valid": {
+			old: batch.Job{
+				ObjectMeta: metav1.ObjectMeta{Name: "abc", Namespace: metav1.NamespaceDefault},
+				Spec: batch.JobSpec{
+					Selector:       validGeneratedSelector,
+					Template:       validPodTemplateSpecForGenerated,
+					Completions:    pointer.Int32Ptr(1),
+					Parallelism:    pointer.Int32Ptr(1),
+					CompletionMode: completionModePtr(batch.IndexedCompletion),
+				},
+			},
+			update: func(job *batch.Job) {
+				job.Spec.Completions = pointer.Int32Ptr(2)
+				job.Spec.Parallelism = pointer.Int32Ptr(2)
+			},
+		},
+		"previous parallelism != previous completions, new parallelism == new completions": {
+			old: batch.Job{
+				ObjectMeta: metav1.ObjectMeta{Name: "abc", Namespace: metav1.NamespaceDefault},
+				Spec: batch.JobSpec{
+					Selector:       validGeneratedSelector,
+					Template:       validPodTemplateSpecForGenerated,
+					Completions:    pointer.Int32Ptr(1),
+					Parallelism:    pointer.Int32Ptr(2),
+					CompletionMode: completionModePtr(batch.IndexedCompletion),
+				},
+			},
+			update: func(job *batch.Job) {
+				job.Spec.Completions = pointer.Int32Ptr(3)
+				job.Spec.Parallelism = pointer.Int32Ptr(3)
+			},
+			err: &field.Error{
+				Type:  field.ErrorTypeInvalid,
+				Field: "spec.completions",
+			},
+		},
+		"indexed job updating completions and parallelism to different values is invalid": {
+			old: batch.Job{
+				ObjectMeta: metav1.ObjectMeta{Name: "abc", Namespace: metav1.NamespaceDefault},
+				Spec: batch.JobSpec{
+					Selector:       validGeneratedSelector,
+					Template:       validPodTemplateSpecForGenerated,
+					Completions:    pointer.Int32Ptr(1),
+					Parallelism:    pointer.Int32Ptr(1),
+					CompletionMode: completionModePtr(batch.IndexedCompletion),
+				},
+			},
+			update: func(job *batch.Job) {
+				job.Spec.Completions = pointer.Int32Ptr(2)
+				job.Spec.Parallelism = pointer.Int32Ptr(3)
+			},
+			err: &field.Error{
+				Type:  field.ErrorTypeInvalid,
+				Field: "spec.completions",
+			},
+		},
+		"indexed job with completions set updated to nil does not panic": {
+			old: batch.Job{
+				ObjectMeta: metav1.ObjectMeta{Name: "abc", Namespace: metav1.NamespaceDefault},
+				Spec: batch.JobSpec{
+					Selector:       validGeneratedSelector,
+					Template:       validPodTemplateSpecForGenerated,
+					Completions:    pointer.Int32Ptr(1),
+					Parallelism:    pointer.Int32Ptr(1),
+					CompletionMode: completionModePtr(batch.IndexedCompletion),
+				},
+			},
+			update: func(job *batch.Job) {
+				job.Spec.Completions = nil
+				job.Spec.Parallelism = pointer.Int32Ptr(3)
+			},
+			err: &field.Error{
+				Type:  field.ErrorTypeRequired,
+				Field: "spec.completions",
+			},
+		},
+		"indexed job with completions unchanged, parallelism reduced to less than completions": {
```

Comment:
> The above behavior can be modeled using an Indexed job with
>
>  - spec.completions equal to spec.parallelism
>  - mutable spec.completions and spec.parallelsim as long as they are mutated together.

As per the above statement in the TEP, the `Completions` and `Parallelism` should be modified together. If only `Parallelism` is modified, should't it throw error? Am I missing something?

Same comment applies for the next test case as well.

Reply:
If the behaviour shown by the test cases `indexed job with completions unchanged, parallelism increased higher than completions` and `indexed job with completions unchanged, parallelism reduced to less than completions` are expected, once `Parallelism` and `Completions` deviate, we cannot mutate `Completions` further.

@ahg-g , please confirm if these test cases are valid.

Reply:
@sathyanarays Parallelism is a mutable value, and in these 2 test cases *only* parallelism is being modified, not completions. Completions is immutable unless completions == parallelism both before and after the update.

Reply:
Let's say initial state is **(C,P)** = **(a,b)** where `a!=b` . 

With this PR, we are restricting the transition **(a,b)** `->` **(c,d)** where `c!=d`.

But, I can achieve this by the following transitions:
**(a,b)** `->` **(a,a)** `->` **(c,c)** `->` **(c,d)**.

It's both theoretically and practically possible to do the above transition so fast that the job controller does reconcile only for the last tuple **(c,d)** ignoring the intermediate states **(b,b)** and **(c,c)**. The effect will be same as the transition **(a,b)** `->` **(c,d)**. Isn't it? So, what's the point in adding this restriction?

@ahg-g , please revert only if my makes sense. 

I would recommend not to waste your time if I am missing something fundamental here. I will figure it out 😆

Reply:
I agree with you, we had a discussion about this here https://github.com/kubernetes/enhancements/pull/3724#discussion_r1086673278, but I didn't push too much, feels minor to me; If in general restricting to before and after makes it more consistent, at least in theory, then may be that is good enough.